models/recurrent_neuron_transformer.py

```diff
@@ -80,314 +78,93 @@ def forward(self, inputs):
-        outputs = self.feedforward_layer(outputs)
-        outputs = self.final_layer(outputs)
+        outputs, hidden_layers = self.multi_head_attention(outputs, hidden_layers)
+        outputs, hidden_layers = self.feedforward_layer(outputs, hidden_layers)
+        outputs, hidden_layers = self.final_layer(outputs, hidden_layers)
         
-        ##############################################################################
-        #                               END OF YOUR CODE                             #
-        ##############################################################################
-        return outputs
+        return outputs, hidden_layers
     
     
     def embed(self, inputs):
         """
         :param inputs: intTensor of shape (N,T)
         :returns embeddings: floatTensor of shape (N,T,H)
         """
-        #############################################################################
-        # TODO:
-        # Deliverable 1: Implement the embedding lookup.                            #
-        # Note: word_to_ix has keys from 0 to self.vocab_size - 1                   #
-        # This will take a few lines.                                               #
-        #############################################################################
       
         embeddings = self.embeddingL(inputs)
         pos_indices = torch.arange(0, inputs.size(1), device=self.device).unsqueeze(0).repeat(inputs.size(0), 1)
         pos_embeddings = self.posembeddingL(pos_indices)
         embeddings  = embeddings + pos_embeddings
-        ##############################################################################
-        #                               END OF YOUR CODE                             #
-        ##############################################################################
+
         return embeddings
         
-    def multi_head_attention(self, inputs):
+    def multi_head_attention(self, inputs, hidden_layers):
         """
         :param inputs: float32 Tensor of shape (N,T,H)
         :returns outputs: float32 Tensor of shape (N,T,H)
         
         Traditionally we'd include a padding mask here, so that pads are ignored.
         This is a simplified implementation.
         """
-        
-        
-        #############################################################################
-        # TODO:
-        # Deliverable 2: Implement multi-head self-attention followed by add + norm.#
-        # Use the provided 'Deliverable 2' layers initialized in the constructor.   #
-        #############################################################################
+        N, T, _ = inputs.shape
+        mask = torch.triu(torch.ones((T, T), device=self.device), diagonal=1).bool()
+
         # Head #1
-        k1 = self.k1(inputs)
-        v1 = self.v1(inputs)
-        q1 = self.q1(inputs)
+        k1, hidden_layers["k1"] = self.k1(inputs, hidden_layers.get("k1"))
```

Comment:
I'm really curious to see how the performance changes when we only replace the final project linear layer with stateful neurons vs. all the KQV matrices. I still need to wrap my head conceptually around what this is going to do, but I think experimenting with it is definitely the right way to proceed, particularly in light of that insightful LinkedIn post you shared on viewing ML as more of an exploratory and experimental "science", rather than purely seeking to confirm some theoretical insight or hunch.